bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitorWorker.java

```diff
@@ -0,0 +1,283 @@
+		 */
+		String manifestToProcessKey = null;
+		ObjectListing objectListing = s3Client.listObjects(bucketListRequest);
+		do {
+			for (S3ObjectSummary objectSummary : objectListing.getObjectSummaries()) {
+				String key = objectSummary.getKey();
+				if (REGEX_MANIFEST.matcher(key).matches()) {
+					/*
+					 * We've got an object that *looks like* it might be a
+					 * manifest file. But we also need to ensure that it starts
+					 * with a valid timestamp, and if so, check to see if that
+					 * timestamp is the oldest one we've encountered so far. If
+					 * so, we mark it as the "current oldest" and continue
+					 * looking through the other keys.
+					 */
+					Instant dataSetTimestamp = parseDataSetTimestamp(key);
+					if (dataSetTimestamp != null && manifestToProcessKey == null)
+						manifestToProcessKey = key;
+					else if (dataSetTimestamp != null && manifestToProcessKey != null
+							&& (dataSetTimestamp.compareTo(parseDataSetTimestamp(manifestToProcessKey)) < 0))
+						manifestToProcessKey = key;
+				}
+			}
+
+			objectListing = s3Client.listNextBatchOfObjects(objectListing);
+		} while (objectListing.isTruncated());
+
+		// If no manifest was found, we're done (until next time).
+		if (manifestToProcessKey == null)
+			return;
+
+		// We've found the oldest manifest. Now go download and parse it.
+		DataSetManifest dataSetManifest = readManifest(manifestToProcessKey);
+
+		/*
+		 * We've got a dataset to process. However, it might still be uploading
+		 * to S3, so we need to wait for that to complete before we start
+		 * processing it.
+		 */
+		boolean dataSetComplete = false;
+		do {
+			if (dataSetIsAvailable(dataSetManifest))
+				dataSetComplete = true;
+
+			/*
+			 * We'll loop forever, but it's prudent to pause between each
+			 * iteration.
+			 */
+			try {
+				Thread.sleep(1000 * 1);
+			} catch (InterruptedException e) {
+				/*
+				 * Many Java applications use InterruptedExceptions to signal
+				 * that a thread should stop what it's doing ASAP. This app
+				 * doesn't, so this is unexpected, and accordingly, we don't
+				 * know what to do. Safest bet is to blow up.
+				 */
+				throw new RuntimeException(e);
+			}
+		} while (!dataSetComplete);
+
+		/*
+		 * Huzzah! We've got a data set to process and we've verified it's all
+		 * there and ready to go. Now we can hand that off to the
+		 * DataSetProcessor, to do the *real* work of actually processing that
+		 * data set. It's important that we block until it's completed, in order
+		 * to ensure that we don't end up processing multiple data sets in
+		 * parallel (which would lead to data consistency problems).
+		 */
+		Set<RifFile> rifFiles = dataSetManifest.getEntries().stream()
+				.map(e -> String.format("%s/%s", DateTimeFormatter.ISO_INSTANT.format(dataSetManifest.getTimestamp()),
+						e.getName()))
+				.map(k -> new S3RifFile(s3Client, new GetObjectRequest(bucketName, k))).collect(Collectors.toSet());
```

Comment:
Does it make sense at this point to get the input stream from the S3 object here, and then you can continue to just use a regular RifFile and not need a S3RifFile. Then the DataSetProcessor doesn't need to know anything about S3.